lib/utils/utils.go

```diff
@@ -56,29 +59,136 @@ func DropInDir(unit string) string {
 // SafeWriteFile is similar to ioutil.WriteFile, but operates by writing to a temporary file first
 // and then relinking the file into the filename which should be an atomic operation. This should be
 // safer, that if the destination file is being replaced, it won't be left in a partial written state.
-func SafeWriteFile(filename string, data []byte, perm os.FileMode) error {
+func SafeWriteFile(filename string, w io.WriterTo, perm os.FileMode) error {
 	dir := filepath.Dir(filename)
 
 	tmpFile, err := ioutil.TempFile(dir, "safewrite")
 	if err != nil {
-		return trace.Wrap(err)
+		return trace.ConvertSystemError(err)
 	}
 	defer os.Remove(tmpFile.Name())
 
-	_, err = tmpFile.Write(data)
+	_, err = w.WriteTo(tmpFile)
 	if err != nil {
-		return trace.Wrap(err)
+		return trace.ConvertSystemError(err)
 	}
 
 	err = os.Chmod(tmpFile.Name(), perm)
 	if err != nil {
-		return trace.Wrap(err)
+		return trace.ConvertSystemError(err)
 	}
 
 	err = os.Rename(tmpFile.Name(), filename)
 	if err != nil {
-		return trace.Wrap(err)
+		return trace.ConvertSystemError(err)
 	}
 
 	return nil
 }
+
+// CopyFile copies contents of src to dst atomically
+// using SharedReadWriteMask as permissions.
+func CopyFile(dst, src string) error {
+	return CopyFileWithPerms(dst, src, constants.SharedReadWriteMask)
+}
+
+// CopyReader copies contents of src to dst atomically
+// using SharedReadWriteMask as permissions.
+func CopyReader(dst string, src io.Reader) error {
+	return CopyReaderWithPerms(dst, src, constants.SharedReadWriteMask)
+}
+
+// CopyFileWithPerms copies the contents from src to dst atomically.
+// Uses CopyReaderWithPerms for its implementation - see function documentation
+// for details of operation
+func CopyFileWithPerms(dst, src string, perm os.FileMode) error {
+	in, err := os.Open(src)
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	defer in.Close()
+	return CopyReaderWithPerms(dst, in, perm)
+}
+
+// CopyReaderWithPerms copies the contents from src to dst atomically.
+// If dst does not exist, CopyReaderWithPerms creates it with permissions perm.
+// If the copy fails, CopyReaderWithPerms aborts and dst is preserved.
+// Adopted with modifications from https://go-review.googlesource.com/#/c/1591/9/src/io/ioutil/ioutil.go
+func CopyReaderWithPerms(dst string, src io.Reader, perm os.FileMode) error {
+	tmp, err := ioutil.TempFile(filepath.Dir(dst), "")
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	cleanup := func() error {
+		err := os.Remove(tmp.Name())
+		if err != nil {
+			log.WithFields(log.Fields{
+				"error": err,
+				"file":  tmp.Name(),
+			}).Warn("Failed to remove.")
+		}
+		return trace.ConvertSystemError(err)
```

Comment:
The cleanup's return argument is not actually used anywhere in this function so it can very well be made niladic instead.